src/api.py

```diff
@@ -33,40 +33,64 @@
     retry_if_exception_type,
     stop_after_attempt,
     before_sleep_log,
-    wait_exponential_jitter,
+    wait_exponential_jitter, retry_if_exception,
 )
 
-VALID_MODELS_FOR_BILLING = [
-    "gpt-4",
-    "gpt-4-0314",
-    "gpt-4-0613",
-    "gpt-4-32k",
-    "gpt-4-32k-0613",
-    "gpt-3.5-turbo",
-    "gpt-3.5-turbo-0613",
-    "gpt-3.5-turbo-16k",
-    "gpt-3.5-turbo-16k-0613",
-    "gpt-4-1106-preview"
-]
-
-
-class GPT4Plugin(StreamingGenerator):
+
+class BillingCallback:
+    def __init__(self):
+        self.cost = None
+        self.completion_response = None
+
+    def __call__(self, kwargs, completion_response, start_time, end_time):
+        if "complete_streaming_response" in kwargs:
+            completion_response = kwargs["complete_streaming_response"]
+            self.cost = litellm.completion_cost(completion_response)
+            self.completion_response = completion_response
+
+    def usage(self, completion_id: str) -> [UsageReport]:
+        while self.cost is None:
+            pass
+        return [
+            UsageReport(
+                operation_type=OperationType.RUN,
+                operation_unit=OperationUnit.UNITS,
+                operation_amount=int(self.cost * 10_000_000_000_000),
```

Comment:
This is large but some of the costs based on what I can find in their calculations get pretty small, and this is the best way I can figure not to wildly over/undercharge in those cases.

Reply:
Python integers are arbitrary-precision, so no.